File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case07_processfun/ProcessFun.java
package com.yw.flink.example.javacases.case07_processfun;

import com.yw.flink.example.StationLog;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

/**
 * Flink读取Socket中通话数据，如果被叫手机连续5s呼叫失败生成告警信息。
 * <p>
 * 002,187,186,fail,2000,20
 * 002,187,188,success,2000,20
 */
public class ProcessFun {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //002,187,186,fail,2000,20
        DataStreamSource<String> ds = env.socketTextStream("node5", 9999);

        //对数据进行转换
        SingleOutputStreamOperator<StationLog> ds2 = ds.map(new MapFunction<String, StationLog>() {
            @Override
            public StationLog map(String s) throws Exception {
                String[] split = s.split(",");
                return new StationLog(split[0], split[1], split[2], split[3], Long.valueOf(split[4]), Long.valueOf(split[5]));
            }
        });

        //对被叫手机进行分组
        KeyedStream<StationLog, String> keyedStream = ds2.keyBy(new KeySelector<StationLog, String>() {
            @Override
            public String getKey(StationLog stationLog) throws Exception {
                return stationLog.callIn;
            }
        });

        //使用processfunction 进行处理，针对每个被叫手机注册定时器
        SingleOutputStreamOperator<String> result = keyedStream.process(new KeyedProcessFunction<String, StationLog, String>() {
            //定义Flink 状态 - 保存触发器触发的时间
            ValueState<Long> timeState = null;

            //在open 方法中对时间状态初始化
            @Override
            public void open(Configuration parameters) throws Exception {
                ValueStateDescriptor<Long> descriptor = new ValueStateDescriptor<>("time", Long.class);
                timeState = getRuntimeContext().getState(descriptor);
            }

            //每条数据进来后都会调用
            @Override
            public void processElement(StationLog stationLog, KeyedProcessFunction<String, StationLog, String>.Context ctx, Collector<String> collector) throws Exception {
                //002,187,186,fail,2000,20
                //002,187,188,success,2000,20
                //获取该key(被叫手机)定时器触发的时间状态
                Long time = timeState.value();

                if ("fail".equals(stationLog.callType) && time == null) {
                    //数据通话失败，需要注册定时器
                    //获取当前时间
                    long nowTime = ctx.timerService().currentProcessingTime();
                    //设置定时器触发时间，5秒后触发
                    long triggerTime = nowTime + 5000L;
                    //注册定时器
                    ctx.timerService().registerProcessingTimeTimer(triggerTime);

                    //对状态中保存的触发器时间进行更新
                    timeState.update(triggerTime);
                }

                if (!"fail".equals(stationLog.callType) && time != null) {
                    //在定时器触发之前，如果该被叫手机号 呼叫非失败，要取消定时器
                    ctx.timerService().deleteProcessingTimeTimer(time);
                    //清除状态值
                    timeState.clear();
                }


            }

            //定时器触发方法
            @Override
            public void onTimer(long timestamp, KeyedProcessFunction<String, StationLog, String>.OnTimerContext ctx, Collector<String> out) throws Exception {
                out.collect("触发时间 = " + timestamp + ",被叫手机号：" + ctx.getCurrentKey() + "连续5秒呼叫失败！！！");

                //清空定时器
                timeState.clear();

            }
        });

        result.print();

        env.execute();
    }
}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case07_processfun/ProcessFun.scala
package com.yw.flink.example.scalacases.case07_processfun

import com.yw.flink.example.StationLog
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.util.Collector

/**
 * 案例：Flink读取Socket中通话数据，如果被叫手机连续5s呼叫失败生成告警信息。
 */
object ProcessFun {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    import org.apache.flink.streaming.api.scala._

    val ds: DataStream[String] = env.socketTextStream("node5", 9999)

    val ds2: DataStream[StationLog] = ds.map(one => {
      val split: Array[String] = one.split(",")
      StationLog(split(0), split(1), split(2), split(3), split(4).toLong, split(5).toLong)
    })

    //按照被叫手机号分组
    val keyedStream: KeyedStream[StationLog, String] = ds2.keyBy(stationLog => {
      stationLog.callIn
    })

    //使用processFunction 进行定时器注册和触发
    keyedStream.process(new KeyedProcessFunction[String,StationLog,String] {
      //定义一个状态 ，记录当前被叫手机号定时器触发时间
      lazy val timeState: ValueState[Long] = getRuntimeContext.getState(new ValueStateDescriptor[Long]("time", classOf[Long]))

      //来一条数据处理一次
      override def processElement(stationLog: StationLog,
                                  ctx: KeyedProcessFunction[String, StationLog, String]#Context,
                                  collector: Collector[String]): Unit = {
        //获取当前被叫手机号的定时器触发时间
        val time: Long = timeState.value()

        //如果来的这条通话数据通话状态为 fail ,并且状态中没有值，注册定时器
        if("fail".equals(stationLog.callType) && time == 0){
          //获取当前时间
          val nowTime: Long = ctx.timerService().currentProcessingTime()
          //定时器触发时间
          val triggerTime = nowTime + 5000L
          //注册定时器
          ctx.timerService().registerProcessingTimeTimer(triggerTime)
          //更新状态
          timeState.update(triggerTime)
        }

        if(!"fail".equals(stationLog.callType) && time != 0){
          //删除定时器
          ctx.timerService().deleteProcessingTimeTimer(time)

          //清空状态
          timeState.clear()
        }


      }

      /**
       * 定时器触发方法
       */
      override def onTimer(timestamp: Long,
                           ctx: KeyedProcessFunction[String, StationLog, String]#OnTimerContext,
                           out: Collector[String]): Unit ={
        out.collect("当前触发时间："+timestamp+",手机号："+ctx.getCurrentKey+"连续5秒呼叫失败！！！")
        //清空状态
        timeState.clear()
      }

    }).print()

    env.execute()

  }

}
